qcodes/dataset/plotting.py

```diff
@@ -574,36 +580,40 @@ def plot_on_a_plain_grid(x: np.ndarray,
         z_strings = np.unique(z)
         z = _strings_as_ints(z)
 
-    xrow, yrow, z_to_plot = reshape_2D_data(x, y, z)
-
-    # we use a general edge calculator,
-    # in the case of non-equidistantly spaced data
-    # TODO: is this appropriate for a log ax?
-    dxs = np.diff(xrow)/2
-    dys = np.diff(yrow)/2
-    x_edges = np.concatenate((np.array([xrow[0] - dxs[0]]),
-                              xrow[:-1] + dxs,
-                              np.array([xrow[-1] + dxs[-1]])))
-    y_edges = np.concatenate((np.array([yrow[0] - dys[0]]),
-                              yrow[:-1] + dys,
-                              np.array([yrow[-1] + dys[-1]])))
+    if x.ndim == 2 and y.ndim == 2 and z.ndim == 2:
+        if np.logical_or(np.any(np.isnan(x)), np.any(np.isnan(y))):
+            x_to_plot, y_to_plot, z_to_plot = _clip_nan_from_shaped_data(x, y, z)
+            num_points = x_to_plot.size * y_to_plot.size
+        else:
+            # data is on a grid that may or may not be
+            # rectilinear. Rely on matplotlib to plot
+            # this directly
+            x_to_plot, y_to_plot, z_to_plot = x, y, z
```

Comment:
not a critical sugegstion, but i think it would help the readability to change the order of this if-else, so that the if block gets the easy case, and the else block gets the has-nan-somewherei-in-x-or-y case (which in turn inside a helper function split into rectilinear case that is easier, and non-reclilinear case that gets the same treatment as the else clause here)